pkg/daemon/certificate_writer.go

```diff
@@ -102,37 +118,71 @@ func (dn *Daemon) syncControllerConfigHandler(key string) error {
 
 		kubeAPIServerServingCABytes := controllerConfig.Spec.KubeAPIServerServingCAData
 		cloudCA := controllerConfig.Spec.CloudProviderCAData
 		pathToData[caBundleFilePath] = kubeAPIServerServingCABytes
 		pathToData[cloudCABundleFilePath] = cloudCA
+		var data []byte
+		var err error
 
-		for bundle, data := range pathToData {
-			if !strings.HasSuffix(string(data), "\n") {
-				bString := string(data) + "\n"
-				data = []byte(bString)
-			}
-			if Finfo, err := os.Stat(bundle); err == nil {
-				var mode os.FileMode
-				Dinfo, err := os.Stat(filepath.Dir(bundle))
+		cm, cmErr := dn.kubeClient.CoreV1().ConfigMaps("openshift-machine-config-operator").Get(context.TODO(), "kubeconfig-data", v1.GetOptions{})
+		if cmErr != nil {
+			klog.Errorf("Error retrieving kubeconfig-data. %v", cmErr)
+		} else {
+			data, err = cmToData(cm, "ca-bundle.crt")
+			if err != nil {
+				klog.Errorf("kubeconfig-data ConfigMap not populated yet. %v", err)
+			} else if data != nil {
+				currentKC := clientcmdv1.Config{}
+				kcBytes, err := os.ReadFile(kubeConfigPath)
 				if err != nil {
-					mode = defaultDirectoryPermissions
-				} else {
-					mode = Dinfo.Mode()
-				}
-				// we need to make sure we honor the mode of that file
-				if err := writeFileAtomically(bundle, data, mode, Finfo.Mode(), -1, -1); err != nil {
 					return err
 				}
-			} else {
-				if err := writeFileAtomicallyWithDefaults(bundle, data); err != nil {
-					return err
+				if kcBytes != nil {
+					err = yaml.Unmarshal(kcBytes, &currentKC)
+					if err != nil {
+						return fmt.Errorf("Could not unmarshal kubeconfig into struct. Data: %s, Error: %v", string(kcBytes), err)
+					}
+					var newData []byte
+					if len(currentKC.Clusters) > 0 {
+						currentKC.Clusters[0].Cluster.CertificateAuthorityData = data
+						newData, err = json.Marshal(currentKC)
+						if err != nil {
+							return fmt.Errorf("Could not marshal kubeconfig into bytes. Error: %v", err)
+						}
+						oldCMBin, err := json.Marshal(cm)
+						if err != nil {
+							return fmt.Errorf("Could not marshal old configmap %v", err)
+						}
+						newCM := cm
+						newCM.BinaryData["kubeconfig"], err = json.Marshal(currentKC)
+						if err != nil {
+							return fmt.Errorf("Could not marshal new kubeconfig %v", err)
+						}
+						newCMBin, err := json.Marshal(newCM)
+						if err != nil {
+							return fmt.Errorf("Could not marshal new configmap %v", err)
+						}
+						patch, err := jsonmergepatch.CreateThreeWayJSONMergePatch(oldCMBin, newCMBin, oldCMBin)
+						if err != nil {
+							return fmt.Errorf("Could not create patch for  kubeconfig-data %v", err)
+						}
+						_, err = dn.kubeClient.CoreV1().ConfigMaps("openshift-machine-config-operator").Patch(context.TODO(), "kubeconfig-data", types.MergePatchType, patch, v1.PatchOptions{})
+						if err != nil {
+							return fmt.Errorf("Could not patch existing kubeconfig-data %v", err)
+						}
+
+					}
+					pathToData[kubeConfigPath] = newData
```

Comment:
So I managed to get this to somewhat work that I can gather the data. It seems that the original and kubeconfig written by this code isn't the same, but they actually end up generating the same kubeconfig once kubelet is restarted:

Original:
```
{
  "preferences": {},
  "clusters": [
    {
      "name": "local",
      "cluster": {
        "server": "https://api-int.ci-ln-m67xb02-76ef8.aws-2.ci.openshift.org:6443",
        "certificate-authority-data": "..."
      }
    }
  ],
  "users": [
    {
      "name": "kubelet",
      "user": {
        "token": "..."
      }
    }
  ],
  "contexts": [
    {
      "name": "kubelet",
      "context": {
        "cluster": "local",
        "user": "kubelet"
      }
    }
  ],
  "current-context": "kubelet"
}
```

And after you write:
```
{
  "preferences": {},
  "clusters": [
    {
      "name": "ci-ln-m67xb02-76ef8",
      "cluster": {
        "server": "https://api-int.ci-ln-m67xb02-76ef8.aws-2.ci.openshift.org:6443",
        "certificate-authority-data": "..."
      }
    }
  ],
  "users": [
    {
      "name": "kubelet",
      "user": {
        "client-certificate-data": "...",
        "client-key-data": "..."
      }
    }
  ],
  "contexts": [
    {
      "name": "kubelet",
      "context": {
        "cluster": "ci-ln-m67xb02-76ef8",
        "user": "kubelet"
      }
    }
  ],
  "current-context": "kubelet"
}
```

So there is some additional data here BUT I am not sure if any part of that is causing problems.

I did try to manually put back the old kubeconfig data, delete /var/lib... and restart kubelet, and it seems to have generated the same /var/lib/kubelet/kubeconfig file:

```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: ...
    server: https://api-int.ci-ln-m67xb02-76ef8.aws-2.ci.openshift.org:6443
  name: default-cluster
contexts:
- context:
    cluster: default-cluster
    namespace: default
    user: default-auth
  name: default-context
current-context: default-context
kind: Config
preferences: {}
users:
- name: default-auth
  user:
    client-certificate: /var/lib/kubelet/pki/kubelet-client-current.pem
    client-key: /var/lib/kubelet/pki/kubelet-client-current.pem
 ```

so... maybe it's not that? right now the only thing that is broken, as far as I can tell, is the operator pod, and hand restarting that doesn't seem to make a difference... is it possible that its silently failing somewhere? Seems unlikely as well, so I'm a bit stumped.